x-pack/platform/plugins/shared/observability_ai_assistant/server/service/knowledge_base_service/kb_query_to_dsl.ts

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { QueryDslQueryContainer } from '@elastic/elasticsearch/lib/api/types';
+import { KnowledgeBaseQueryContainer } from '.';
+import { ML_INFERENCE_PREFIX } from './recall_from_search_connectors';
+
+function wrapInShould(query: QueryDslQueryContainer[]): QueryDslQueryContainer {
+  return {
+    bool: {
+      should: query,
+      minimum_should_match: 1,
+    },
+  };
+}
+
+export function kbQueryToDsl(
+  query: KnowledgeBaseQueryContainer,
+  field: string,
+  modelId: string
+): QueryDslQueryContainer;
+
+export function kbQueryToDsl(
+  query: KnowledgeBaseQueryContainer,
+  semanticTextFields: string[]
+): QueryDslQueryContainer;
+
+export function kbQueryToDsl(
+  query: KnowledgeBaseQueryContainer,
+  ...rest: [string[]] | [string, string]
+): QueryDslQueryContainer {
```

Comment:
What about creating two separate functions instead of overloading this one? It's more or less two entirely separate code branches, and as it is now, it's quite hard to follow exactly when each branch is met.

Reply:
Perhaps something like this (completely untested).

```ts
export function kbQueryToDsl(
  query: KnowledgeBaseQueryContainer,
  fieldOrFields: string | string[],
  modelId?: string
): QueryDslQueryContainer {
  const fields = Array.isArray(fieldOrFields) ? fieldOrFields : [fieldOrFields];

  if ('keyword' in query) {
    return wrapInShould(
      query.keyword.value.map((keyword) => ({
        multi_match: {
          fields: ['*'],
          query: keyword,
          boost: query.keyword.boost ?? 1,
        },
      }))
    );
  }

  if (modelId) {
    return wrapInShould(fields.map((field) => buildVectorQuery(query, field, modelId)));
  }

  return wrapInShould(fields.map((field) => buildSemanticQuery(query, field)));
}
```